### PR TITLE
Support both tokio and futures, via Cargo features

### PR DESCRIPTION
**Pull request by joshtriplett**
Make tokio dependency optional via a "tokio" feature, and add support for
futures via a "futures" feature.

This attempts to duplicate as little code as possible.

As implemented, exactly one of "futures" and "tokio" must be enabled; enabling
both produces an error due to conflicting implementations of the same traits.
Supporting both simultaneously would require substantially more duplication,
and two distinct families of types.

### CODE REVIEW

**Comment by jonhoo** [comment]
Thanks for digging into this!

I think mutually exclusive features is a non-starter, unfortunately, since it's too likely to cause unresolveable issues for users just by virtue of how Cargo auto-merges features across a crate's dependency closure. It sucks that it _only_ happens because the same `Stream` trait is used in both cases. That said, it _should_ be possible to just have _one_ impl, and then for the client to use an adapter type just for the `R` in the non-tokio case that adapts from impls of `futures-io::AsyncRead` to `tokio::io::AsyncRead`. It's not quite as ergonomic, but at least it makes the rest of the change not require the features to be mutually exclusive.

Barring that, I think the alternative is to split the implementation up further into multiple types. And maybe macros can help us make it so that little code is actually _manually_ repeated. If we can't easily come up with such a macro, I wonder at that point if it even makes sense for those two impls to share a crate in the first place.

A standardized `AsyncRead`/`AsyncWrite` really can't come soon enough :sweat_smile:

**Comment by joshtriplett** [comment]
@jonhoo Adapting from one read trait to the other would require a dependency on both, rather than being able to feature-flag away the other dependency. However, if you don't mind adapter types, I do have an idea that'll allow both impls to coexist.

**Comment by jonhoo** [comment]
I'm not too concerned with the `futures` feature also pulling in `tokio`, as long as it does so with basically no features enabled, since then very little is compiled anyway.

**Comment by joshtriplett** [comment]
@jonhoo I am. I think I can make the two features completely independent without too much difficulty, using a wrapper type for the stream impl.

File: src/futures.rs
Comment (jonhoo) [inline src/futures.rs:3]: This doesn't seem right. There is no top-level version of `AsyncBincodeReader` or any of the other types any more. I think we should probably place that documentation directly into the generated per-module types.

Comment (jonhoo) [reply src/futures.rs:3]: Same goes for the analogous comment in `mod tokio`.

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:17]: We should document here that all the relevant types are in the sub-modules, and direct people towards the appropriate one for their use-case.

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:21]: To avoid making this a backwards-incompatible change, I'd like to re-export the types from `mod tokio` here and do a point release with that in place.

File: src/reader.rs
Comment (jonhoo) [inline src/reader.rs:12]: The more I look at this macro, the more I feel like it _should_ be doable without a macro. I tried mocking something up on the playground and it _seems_ to work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2d284f511e2fa13170515ed6bd6f373d

What do you think of that approach instead? Is it something you've tried and that turns out not to work (due to GATs or something along those lines)?

**Review by jonhoo — CHANGES_REQUESTED** [review]
I'm happy we're able to do a fair amount of code sharing with this new approach! I left some inline notes, the biggest of which is a proposal to do this without a macro. I don't know if it's actually tenable (or if you've tried it yourself already), but I think it _could_ make the change even nicer!

File: src/futures.rs
Comment (joshtriplett) [reply src/futures.rs:3]: "top-level documentation *and* the documentation for" - the macros generate these types, and they exist in the same module as this comment, which is where the links go.

File: src/lib.rs
Comment (joshtriplett) [reply src/lib.rs:21]: Sure, easy to do.

File: src/reader.rs
Comment (joshtriplett) [reply src/reader.rs:12]: @jonhoo It'd be fairly straightforward to do that for the reader code; I did go down a similar route at first. It seems much harder to get that working for the writer and stream code; even though the traits are similar for the two frameworks, a fair bit of similar code would still get duplicated twice.

I think it's *absolutely* possible, but I don't think it's going to be elegant.

I also don't think this would be as maintainable (though as the maintainer that's your call).

And, apart from that, I'm also hoping the half-life of this division doesn't last much longer; I hope version 0.8 (or 1.0) of this crate uses traits from std, and near-zero async-framework-specific code.

File: src/lib.rs
Comment (joshtriplett) [reply src/lib.rs:17]: Will do.

File: src/futures.rs
Comment (joshtriplett) [reply src/futures.rs:3]: I think I've resolved this by expanding the documentation, but let me know what you think.

File: src/reader.rs
Comment (jonhoo) [reply src/reader.rs:12]: Fair enough, I trust your assessment here :+1: 
I was thinking doing it with another generic argument might make it easier for _consumers_ of this crate to also be framework agnostic, but maybe that's not true either.

Let's land this as-is, and then we can always iterate further down the line should it be necessary.

And yes, can't wait for shared traits!

**Review by jonhoo — APPROVED** [review]

**Comment by jonhoo** [comment]
Released as 0.6.2 :tada:

File: src/reader.rs
Comment (joshtriplett) [reply src/reader.rs:12]: > I was thinking doing it with another generic argument might make it easier for _consumers_ of this crate to also be framework agnostic, but maybe that's not true either.

Ah, I see! Yeah, I think that's theoretically possible, but it'd be a much larger engineering challenge to make that portability layer useful for other code, and I think that's a separate project.

(Also, a couple of portability layers already exist, and people tend to not want to use them because they're an extra layer/crate.)

**Pull request merged**
